# Set initialization file
https://cccma.gitlab.io/classic/makeInputFiles.html

Initialize the model for CLASSIC (CLASS+CTEM) run. Use the Paul's SnowMIP data to set the first soil layers proporties
and then extrapolate with the satellite data. 

In [1]:
# Env: sc2_v0

import xarray as xr
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import proplot as pplt # New plot library (https://proplot.readthedocs.io/en/latest/)
pplt.rc['savefig.dpi'] = 300 # 1200 is too big! #https://proplot.readthedocs.io/en/latest/basics.html#Creating-figures

In [2]:
exp = 'Ref'
site = 'obs'
site_paul = 'Old_Black_Spruce'
site_ex = 'CA-Obs'
# site_ex = 'AU-Tum'

path_in = '/home/lalandmi/eccc/classic-develop/inputFiles/FLUXNETsites/'+site_ex # example file
path_out = '/home/lalandmi/eccc/classic-develop/inputFiles/SnowMIP/'+site

In [3]:
ds = xr.open_dataset(path_in+'/'+site_ex+'_init.nc')
ds.load()

<xarray.Dataset>
Dimensions:            (tile: 1, lat: 1, lon: 1, icp1: 6, layer: 20, ic: 5,
                        icc: 12, iccp1: 13, iccp2: 14, slope: 8)
Coordinates:
  * ic                 (ic) int32 1 2 3 4 5
  * icc                (icc) int32 1 2 3 4 5 6 7 8 9 10 11 12
  * iccp1              (iccp1) int32 1 2 3 4 5 6 7 8 9 10 11 12 13
  * iccp2              (iccp2) int32 1 2 3 4 5 6 7 8 9 10 11 12 13 14
  * icp1               (icp1) int32 1 2 3 4 5 6
  * lat                (lat) float64 53.99
  * layer              (layer) int32 1 2 3 4 5 6 7 8 ... 13 14 15 16 17 18 19 20
  * lon                (lon) float64 -105.1
  * slope              (slope) int32 1 2 3 4 5 6 7 8
  * tile               (tile) int32 1
Data variables: (12/100)
    ALBS               (tile, lat, lon) float64 0.8
    ALIC               (tile, icp1, lat, lon) float64 0.19 0.0 0.0 0.0 0.0 0.0
    ALVC               (tile, icp1, lat, lon) float64 0.03 0.0 0.0 0.0 0.0 0.0
    CLAY               (tile, layer, lat, lon) float64 0.0 0.0 0.0 ... 1.6 0.0
    CMAS               (tile, ic, lat, lon) float64 4.8 0.0 0.0 0.0 0.0
    Cmossmas           (tile, lat, lon) float64 0.0
    ...                 ...
    stemmass_s         (tile, icc, lat, lon) float64 4.5 4.5 0.0 ... 0.0 0.0 0.0
    stmhrlos           (tile, icc, lat, lon) float64 0.0 0.0 0.0 ... 0.0 0.0 0.0
    tcoldm             (lat, lon) float64 99.0
    twarmm             (lat, lon) float64 -99.0
    tymaxlai           (tile, icc, lat, lon) float64 0.0 0.0 0.0 ... 0.0 0.0 0.0
    nh4_mass           (tile, iccp1, lat, lon) float32 0.0 0.0 0.0 ... 0.0 0.0
Attributes:
    title:          CLASSIC initialization file
    creation_date:  20220428

In [4]:
ds.soilcmas.sum()

<xarray.DataArray 'soilcmas' ()>
array(0.)

## BERMS Old Black Spruce, Saskatchewan, Canada

Predominantly black spruce, with some tamarack, jack pine and balsam poplar, naturally
regenerated after fire in 1894. Sparse understorey. Peat soil over sandy loam and sand
with a raised water table.

| Short name | obs |
|:-----------|:----|
| Location | 53.99ºN, 105.12ºW |
| Elevation | 629 m |
| Canopy height | 11 m |
| Leaf area index | 3.5 – 3.8 |
| Snow-free albedo (above canopy) | 0.08 |
| Simulation period | 1 October 1997 to 30 September 2010 |
| Temperature/humidity measurement height | 25 m |
| Wind measurement height | 26 m |
| Reference | Bartlett et al. (2007) |

In [3]:
lat = 53.98717
lon = -105.11779

# If the site description is not enough we can get the input data from gridded satellite datasets
path = '/home/lalandmi/Dropbox/data/CLASSIC/'
SoilGrids250m_CLAY = xr.open_dataset(path+'/soil/SoilGrids250m_CLAY_0.05deg.nc')
SoilGrids250m_ORGM = xr.open_dataset(path+'/soil/SoilGrids250m_ORGM_0.05deg.nc')
SoilGrids250m_SAND = xr.open_dataset(path+'/soil/SoilGrids250m_SAND_0.05deg.nc')
SoilGrids250m_SDEP = xr.open_dataset(path+'/soil/SoilGrids250m_SDEP_0.05deg.nc')
NCAR_SOCI = xr.open_dataset(path+'/soil/NCAR_SOCI_0.5deg.nc')
c4_fraction_1deg = xr.open_dataarray(path+'/vegetation/c4_fraction_1deg.nc')

path_paul = '/home/lalandmi/Dropbox/data/SnowMIP/Paul' 
ds_init_Paul = xr.open_dataset(path_paul+'/'+site_paul+'_60_ESMSnowMIP.nc').load()
ds_init_Paul

<xarray.Dataset>
Dimensions:            (lon: 1, lat: 1, tile: 1, icp1: 5, layer: 3, ic: 4,
                        icctem: 9, iccp2: 11, slope: 8)
Coordinates:
  * lon                (lon) float64 254.9
  * lat                (lat) float64 53.99
  * tile               (tile) int32 1
  * icp1               (icp1) int32 1 2 3 4 5
  * layer              (layer) int32 1 2 3
  * ic                 (ic) int32 1 2 3 4
  * icctem             (icctem) int32 1 2 3 4 5 6 7 8 9
  * iccp2              (iccp2) int32 1 2 3 4 5 6 7 8 9 10 11
  * slope              (slope) int32 1 2 3 4 5 6 7 8
Data variables: (12/43)
    GC                 (lat, lon) float64 -1.0
    nmtest             (lat, lon) float64 1.0
    FCAN               (tile, icp1, lat, lon) float64 1.0 0.0 0.0 0.0 0.0
    LNZ0               (tile, icp1, lat, lon) float64 0.095 0.0 0.0 0.0 0.0
    ALIC               (tile, icp1, lat, lon) float64 0.16 0.0 0.0 0.0 0.0
    ALVC               (tile, icp1, lat, lon) float64 0.03 0.0 0.0 0.0 0.0
    ...                 ...
    srplsmon           (lat, lon) float64 8.198e-319
    defctmon           (lat, lon) float64 7.996e-319
    anndefct           (lat, lon) float64 9.321e-314
    annsrpls           (lat, lon) float64 7.905e-323
    annpcp             (lat, lon) float64 0.0
    dry_season_length  (lat, lon) float64 0.0
Attributes:
    title:          CLASSIC initialization file created from: /home/lalandmi/...
    creation_date:  20240123

### Set the layer coords to the center soil layer

In [6]:
layer_c = []

for i in range(len(ds.DELZ)):
    if i == 0:
        layer_c.append(ds.DELZ.cumsum().values[i]/2)
    else:
        layer_c.append(ds.DELZ.cumsum().values[i-1] + ds.DELZ.values[i]/2)
        
with xr.set_options(keep_attrs=True):
    ds = ds.assign_coords(layer=ds.layer*0+layer_c)

ds.layer

<xarray.DataArray 'layer' (layer: 20)>
array([5.00e-02, 1.50e-01, 2.50e-01, 3.50e-01, 4.50e-01, 5.50e-01, 6.50e-01,
       7.50e-01, 8.50e-01, 9.50e-01, 1.10e+00, 1.35e+00, 1.70e+00, 2.15e+00,
       2.90e+00, 4.90e+00, 8.90e+00, 1.89e+01, 4.14e+01, 5.89e+01])
Coordinates:
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9
Attributes:
    standard_name:  ground layers

In [7]:
ds.DELZ.cumsum()

<xarray.DataArray 'DELZ' (layer: 20)>
array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.2,
        1.5,  1.9,  2.4,  3.4,  6.4, 11.4, 26.4, 56.4, 61.4])
Coordinates:
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9

In [8]:
layer_c = []

for i in range(len(ds_init_Paul.DELZ)):
    if i == 0:
        layer_c.append(ds_init_Paul.DELZ.cumsum().values[i]/2)
    else:
        layer_c.append(ds_init_Paul.DELZ.cumsum().values[i-1] + ds_init_Paul.DELZ.values[i]/2)
        
with xr.set_options(keep_attrs=True):
    ds_init_Paul = ds_init_Paul.assign_coords(layer=ds_init_Paul.layer*0+layer_c)

ds_init_Paul.layer

<xarray.DataArray 'layer' (layer: 3)>
array([0.05 , 0.225, 2.225])
Coordinates:
  * layer    (layer) float64 0.05 0.225 2.225
Attributes:
    standard_name:  ground layers

In [9]:
ds_init_Paul.DELZ.cumsum()

<xarray.DataArray 'DELZ' (layer: 3)>
array([0.1 , 0.35, 4.1 ])
Coordinates:
  * layer    (layer) float64 0.05 0.225 2.225

### Combine Paul's data with satellite data

#### SAND

In [10]:
ds_init_Paul.SAND.squeeze().values

array([-2. , -2. , 95.5])

In [13]:
ds.SAND.squeeze().values # values set at the site so keep them as is

array([-2. , -2. , -2. , 72.7, 64.5, 96. , 96. , 95.5, 95.5, 95.5, 95.5,
       95.5, 95.5, 95.5, 95.5, 95.5, 95.5, 95.5, 95.5, -3. ])

In [11]:
SoilGrids250m_SAND.sel(lat=lat, lon=lon, method='nearest').SAND.values

array([28.826199, 36.298817, 39.975098, 42.446205, 42.446205, 42.446205,
       43.882423, 43.882423, 43.882423, 43.882423, 43.919514, 43.919514,
       43.919514, 43.919514, 43.919514, 43.919514, 43.919514, 43.919514,
       43.919514, 43.919514], dtype=float32)

In [14]:
# SAND = ds_init_Paul.where(ds_init_Paul.SAND >= 0).interp(layer=ds.layer, method='linear').SAND.squeeze()
# SAND = SAND.interpolate_na(dim='layer', method='nearest', fill_value="extrapolate")

# # Deals only with the case of the first layer is peat
# # (needs to be adapted if more peat layers)
# if ds_init_Paul.SAND.squeeze().values[0] == -2:
#     SAND.values[0] = -2 

# # Average the last layers with satellite data
# SAND.values[4:] = (SAND.values[4:] + SoilGrids250m_SAND.sel(lat=lat, lon=lon, method='nearest').SAND.values[4:])/2
# SAND

#### CLAY

In [15]:
ds_init_Paul.CLAY.squeeze().values

array([0. , 0. , 1.6])

In [16]:
ds.CLAY.squeeze().values

array([0. , 0. , 0. , 5.7, 6.7, 2.1, 2.1, 1.6, 1.6, 1.6, 1.6, 1.6, 1.6,
       1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 0. ])

In [17]:
SoilGrids250m_CLAY.sel(lat=lat, lon=lon, method='nearest').CLAY.values

array([14.386506, 19.220959, 21.905172, 23.43465 , 23.43465 , 23.43465 ,
       23.47749 , 23.47749 , 23.47749 , 23.47749 , 23.413265, 23.413265,
       23.413265, 23.413265, 23.413265, 23.413265, 23.413265, 23.413265,
       23.413265, 23.413265], dtype=float32)

In [18]:
# CLAY = ds_init_Paul.where(ds_init_Paul.SAND >= 0).interp(layer=ds.layer, method='linear').CLAY.squeeze()
# CLAY = CLAY.interpolate_na(dim='layer', method='nearest', fill_value="extrapolate")

# # Deals only with the case of the first layer is peat
# # (needs to be adapted if more peat layers)
# if ds_init_Paul.SAND.squeeze().values[0] == -2:
#     CLAY.values[0] = 0

# # Average the last layers with satellite data
# CLAY.values[4:] = (CLAY.values[4:] + SoilGrids250m_CLAY.sel(lat=lat, lon=lon, method='nearest').CLAY.values[4:])/2
# CLAY

#### ORGM

In [19]:
ds_init_Paul.ORGM.squeeze().values

array([86.9, 91.5,  0.2])

In [20]:
ds.ORGM.squeeze().values

array([86.9, 91.4, 90. ,  2. ,  1. ,  0.3,  0.3,  0.3,  0.2,  0.2,  0.2,
        0.2,  0.2,  0.2,  0.2,  0.2,  0.2,  0.2,  0. ,  0. ])

In [21]:
SoilGrids250m_ORGM.sel(lat=lat, lon=lon, method='nearest').ORGM.values

array([36.17816 , 19.633621, 11.494252,  6.882183,  6.882183,  6.882183,
        5.45977 ,  5.45977 ,  5.45977 ,  5.45977 ,  5.718391,  5.718391,
        5.718391,  5.718391,  5.718391,  5.718391,  5.718391,  5.718391,
        5.718391,  5.718391], dtype=float32)

In [22]:
# ORGM = ds_init_Paul.where(ds_init_Paul.SAND >= 0).interp(layer=ds.layer, method='linear').ORGM.squeeze()
# ORGM = ORGM.interpolate_na(dim='layer', method='nearest', fill_value="extrapolate")

# # Deals only with the case of the first layer is peat
# # (needs to be adapted if more peat layers)
# if ds_init_Paul.SAND.squeeze().values[0] == -2:
#     ORGM.values[0] = 0

# # Average the last layers with satellite data
# ORGM.values[4:] = (ORGM.values[4:] + SoilGrids250m_ORGM.sel(lat=lat, lon=lon, method='nearest').ORGM.values[4:])/2
# ORGM

In [23]:
ds.SDEP

<xarray.DataArray 'SDEP' (tile: 1, lat: 1, lon: 1)>
array([[[32.]]])
Coordinates:
  * lat      (lat) float64 53.99
  * lon      (lon) float64 -105.1
  * tile     (tile) int32 1
Attributes:
    units:      m
    long_name:  Soil permeable depth

In [24]:
SoilGrids250m_SDEP.sel(lat=lat, lon=lon, method='nearest').SDEP.values.item(0)

17.84000015258789

In [30]:
ds_init_Paul.FCAN

<xarray.DataArray 'FCAN' (tile: 1, icp1: 5, lat: 1, lon: 1)>
array([[[[1.]],

        [[0.]],

        [[0.]],

        [[0.]],

        [[0.]]]])
Coordinates:
  * lon      (lon) float64 254.9
  * lat      (lat) float64 53.99
  * tile     (tile) int32 1
  * icp1     (icp1) int32 1 2 3 4 5
Attributes:
    units:      -
    long_name:  Annual maximum fractional coverage of modelled area (read in ...

In [28]:
ds.FCAN

<xarray.DataArray 'FCAN' (tile: 1, icp1: 6, lat: 1, lon: 1)>
array([[[[1.]],

        [[0.]],

        [[0.]],

        [[0.]],

        [[0.]],

        [[0.]]]])
Coordinates:
  * icp1     (icp1) int32 1 2 3 4 5 6
  * lat      (lat) float64 53.99
  * lon      (lon) float64 -105.1
  * tile     (tile) int32 1
Attributes:
    units:      -
    long_name:  Annual maximum fractional coverage of modelled area (read in ...

In [29]:
ds.fcancmx

<xarray.DataArray 'fcancmx' (tile: 1, icc: 12, lat: 1, lon: 1)>
array([[[[0.9]],

        [[0.1]],

        [[0. ]],

        [[0. ]],

        [[0. ]],

        [[0. ]],

        [[0. ]],

        [[0. ]],

        [[0. ]],

        [[0. ]],

        [[0. ]],

        [[0. ]]]])
Coordinates:
  * icc      (icc) int32 1 2 3 4 5 6 7 8 9 10 11 12
  * lat      (lat) float64 53.99
  * lon      (lon) float64 -105.1
  * tile     (tile) int32 1
Attributes:
    units:      -
    long_name:  PFT fractional coverage per grid cell

## Copy the init file from FLUXNET


In [32]:
!cp {path_in}/{site_ex}_init.nc {path_out}/{site}_init_spinup_{exp}.nc
!cp {path_in}/{site_ex}_init.nc {path_out}/rsfile_spinup_{exp}.nc

In [34]:
!mkdir -p /home/lalandmi/eccc/classic-develop/outputFiles/SnowMIP/{site}/spinup_{exp}